### PR TITLE
feat(query): Partial support of subqueries in parser and minimal impl…

### PR DESCRIPTION
**Pull request by kvpetrov**
…ementation for instant selectors and functions

**Pull Request checklist**

- [ ] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

**New behavior :**

**BREAKING CHANGES**

If this PR contains a breaking change, please describe the impact and migration
path for existing applications.
If not please remove this section.

Breaking changes may include:
- Any schema changes to any Cassandra tables
- The serialized format for Dataset and Column (see .toString methods)
- Over the wire formats for Akka messages / case classes
- Changes to the HTTP public API
- Changes to query parsing / PromQL parsing

**Other information**:

### CODE REVIEW

File: prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala [hunk 1]
```diff
@@ -562,8 +562,14 @@ class ParserSpec extends AnyFunSpec with Matchers {
...
+      "foo[5m:1m]" -> "PeriodicSeries(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(foo))),List(),Some(300000),None),1524855688000,60000,1524855988000,None)",
+      "rate(foo[3m])[5m:1m]" -> "PeriodicSeriesWithWindowing(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(foo))),List(),Some(180000),None),1524855688000,60000,1524855988000,180000,Rate,false,List(),None)"
+      //"max_over_time(rate(foo[5m])[5m:1m])" -> "" //will fail to parse
+      //"max_over_time(sum(rate(foo[5m]))[5m:1m])" -> "", //will fail to parse
+      //"sum(rate(foo[5m])[5m:1m])" -> "" // fail
```
Comment (vishramachandran) [inline prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala:570]: Can we get a query such as this to pass against parser? This is a minimal test case which makes the feature usable.

`sum(rate(foo[5m])[5m:1m])`

I believe you have made `function` as a subqueryable expression. We should get it to work.

File: prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala [hunk 2]
```diff
@@ -562,8 +562,14 @@ class ParserSpec extends AnyFunSpec with Matchers {
...
+      "foo[5m:1m]" -> "PeriodicSeries(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(foo))),List(),Some(300000),None),1524855688000,60000,1524855988000,None)",
+      "rate(foo[3m])[5m:1m]" -> "PeriodicSeriesWithWindowing(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(foo))),List(),Some(180000),None),1524855688000,60000,1524855988000,180000,Rate,false,List(),None)"
```
Comment (vishramachandran) [inline prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala:567]: There are four kinds of test cases here:
1. subquery expression on raw metric like `foo[5m:1m]`
2. subquery expression on transformed metric like `rate(foo[3m])[5m:1m]`
3. subquery expression with range function `max_over_time(sum(rate(foo[5m]))[5m:1m])`
4. subquery expression with range function, followed by further transformation `sum(rate(foo[5m])[5m:1m])`

Priority of cases is: 3 & 4, followed by 2 & 1

Can we make the parser work for cases 3 & 4 first ?

Comment (vishramachandran) [reply prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala:567]: We should also get the resulting logical plan right in the PR. Let's create the SubqueryRangeFunction logical plan and use it.

**Review by vishramachandran — CHANGES_REQUESTED** [review]
Kier, Let's get the P0 subquery cases working with parser. Thanks!

File: prometheus/src/main/scala/filodb/prometheus/ast/Expressions.scala [hunk 3]
```diff
@@ -137,7 +137,62 @@ trait Expressions extends Aggregates with Functions {
...
+
+      // currently we want to support only subqueries over instant selector like:
+      // foo[5m:1m]
+      // and over time range functions like:
+      // rate(foo[3m])[5m:1m]
+      // eventually, we will start supporting subqueries in more cases
+      subquery match {
+        case ie : InstantExpression => ie.toSeriesPlan(timeParamsToUse);
+        case f : Function => functionLogicalPlan(f, sqcl, timeParamsToUse);
+        case _ => ???
```
Comment (kvpetrov) [inline prometheus/src/main/scala/filodb/prometheus/ast/Expressions.scala:185]: put real explanation here

File: prometheus/src/main/scala/filodb/prometheus/ast/TimeUnits.scala [hunk 4]
```diff
@@ -57,4 +57,9 @@ trait TimeUnits {
 
   case class Offset(duration: Duration)
 
+  case class TimeInterval(duration: Duration) {
+  }
+
+  case class SubqueryClause(interval: Duration, step: Option[Duration]) {
+  }
```
[the remark this answers is not in the document]
Comment (kvpetrov) [reply prometheus/src/main/scala/filodb/prometheus/ast/TimeUnits.scala:64]: fxied

File: prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala [hunk 5]
```diff
@@ -218,17 +218,20 @@ trait Vectors extends Scalars with TimeUnits with Base {
...
-        ApplyInstantFunctionRaw(rs, InstantFunctionId.HistogramBucket, Seq(param))
-      }.getOrElse(rs)
+        // multiply by 1000 to convert unix timestamp in seconds to millis
+        val rs = RawSeries(
+          timeParamToSelector(timeParams),
+          columnFilters,
+          column.toSeq,
+          Some(timeInterval.duration.millis(timeParams.step * 1000)),
+          offset.map(_.millis(timeParams.step * 1000))
+        )
```
[the remark this answers is not in the document]
Comment (kvpetrov) [reply prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala:228]: fixed

File: prometheus/src/main/scala/filodb/prometheus/parse/Parser.scala [hunk 6]
```diff
@@ -194,6 +194,15 @@ trait Unit extends BaseParser {
...
     case d ~ tu => Duration(d.toDouble, tu)
   }
+
+  lazy val subqueryClause: PackratParser[SubqueryClause] = "[" ~ duration ~ ":" ~ duration.? ~ "]" ^^ {
+    case leftBracket ~ timeRange ~ colon ~ step ~ rightBracket => SubqueryClause(timeRange, step)
+  }
+
+  lazy val timeInterval: PackratParser[TimeInterval] = "[" ~ duration ~ "]" ^^ {
+    case leftBracket ~ timeRange ~ rightBracket => TimeInterval(timeRange)
+  }
```
[the remark this answers is not in the document]
Comment (kvpetrov) [reply prometheus/src/main/scala/filodb/prometheus/parse/Parser.scala:204]: fixed

File: prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala (hunk 1 above)
Comment (kvpetrov) [reply prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala:570]: this one does not work yet as sub is not a function but an aggregate. I think I want to do it later, for as long as I can make function work, aggregate should come fairly easy.

File: prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala (hunk 2 above)
Comment (kvpetrov) [reply prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala:567]: 1, 2, 3, 4 without aggregates should work now. The new logical plan is there as well.

**Pull request closed without merging**